tools/wqlat.py

```diff
@@ -0,0 +1,181 @@
+#!/usr/bin/env python
+# @lint-avoid-python-3-compatibility-imports
+#
+# wqlat Summarize kernel workqueue latency as a histogram.
+#           For Linux, uses BCC, eBPF.
+#
+# USAGE: wqlat [-h] [-T] [-N] [-W] [-w WQNAME] [interval] [count]
+#
+# Copyright (c) ping gan.
+# Licensed under the Apache License, Version 2.0 (the "License")
+#
+# 29-Jan-2024   ping gan     Created this.
+
+
+from __future__ import print_function
+from bcc import BPF
+from time import sleep, strftime
+import argparse
+import sys
+
+# arguments
+examples = """examples:
+    ./wqlat                   # summarize workqueue latency as a histogram
+    ./wqlat 1 10              # print 1 second summaries, 10 times
+    ./wqlat -W 1 10           # print 1 second, 10 times per workqueue
+    ./wqlat -NT 1             # 1s summaries, nanoseconds, and timestamps
+    ./wqlat -w nvmet_tcp_wq 1 # 1s summaries for workqueue nvmet_tcp_wq
+"""
+parser = argparse.ArgumentParser(
+    description="Summarize workqueue request latency as histograms.",
+    formatter_class=argparse.RawDescriptionHelpFormatter,
+    epilog=examples)
+parser.add_argument("-T", "--timestamp", action="store_true",
+    help="include timestamp on output")
+parser.add_argument("-N", "--nanoseconds", action="store_true",
+    help="output in nanoseconds")
+parser.add_argument("-W", "--workqueues", action="store_true",
+    help="print a histogram per work queue")
+parser.add_argument("-w", "--wqname", type=str,
+    help="print this workqueue only")
+parser.add_argument("interval", nargs="?", default=99999999,
+    help="output interval, in seconds")
+parser.add_argument("count", nargs="?", default=99999999,
+    help="number of outputs")
+parser.add_argument("--ebpf", action="store_true",
+    help=argparse.SUPPRESS)
+args = parser.parse_args()
+countdown = int(args.count)
+if args.nanoseconds:
+    factor = 1
+    label = "nsecs"
+else:
+    factor = 1000
+    label = "usecs"
+debug = 0
+if args.wqname and len(args.wqname) >= 24:
```

Comment:
In kernel, one of workqueue is "events_freezable_power_efficient". Could you increase the length limit from 24 to say 40? Hopefully verification still okay. If not, we might be able to compare e.g., first batch of 8 bytes, after that doing byte-vs-byte comparison.

Reply:
I'm not sure whether length 40 is okay since kernel's workqueue name length is 24, anyway I will try; but if that doesn't work, can I limit the wqname length to 16? I think length 16 can match most cases.

Reply:
> I'm not sure whether length 40 is okay since kernel's workqueue name length is 24, anyway I will try; but if that doesn't work, can I limit the wqname length to 16? I think length 16 can match most cases.

Sorry, I double checked the kernel code and indeed the workqueue name length is 24.
```
  /* in kernel/workqueue.c */
  WQ_NAME_LEN             = 24
```
So your current change looks good to me.